File: Medium/RottingOranges.java
//https://leetcode.com/problems/rotting-oranges/description/

class Solution {
    public int orangesRotting(int[][] grid) {
        int mins = 0;

        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[i].length; j++){
                if(grid[i][j] == 1){
                    mins = Math.max(mins, bfs(grid, i, j, 0, new boolean[grid.length][grid[0].length]));
                    if(mins == Integer.MAX_VALUE){
                        return -1;
                    }
                }
            }
        }

        return mins;
    }

    public int bfs(int[][] grid, int i, int j, int count, boolean[][] visited) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[i].length || visited[i][j] || grid[i][j] == 0) {
            return Integer.MAX_VALUE;
        } else if (grid[i][j] == 2) {
            return count;
        }
        
        visited[i][j] = true;

        int minCount = Integer.MAX_VALUE;
        minCount = Math.min(minCount, bfs(grid, i + 1, j, count + 1, visited));
        minCount = Math.min(minCount, bfs(grid, i - 1, j, count + 1, visited));
        minCount = Math.min(minCount, bfs(grid, i, j + 1, count + 1, visited));
        minCount = Math.min(minCount, bfs(grid, i, j - 1, count + 1, visited));

        visited[i][j] = false;

        return minCount;
    }
}

File: Medium/RottingOranges.py
class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        count = 0
        newCount = 0

        while newCount==0 or newCount != count:
            count = newCount
            newCount = self.update(grid, count)
            
        if 1 in grid:
            return -1
        return newCount
        
    def update(grid, count):
        change = False
        for i in range(0,len(grid)):
            for j in range(0,len(grid[i])):
                    if grid[i][j] == 2 and i > 0 and grid[i-1][j]==1:
                        grid[i-1][j] = 2
                        change = True
                    if grid[i][j] == 2 and i < len(grid)-1 and grid[i+1][j]==1:
                        grid[i+1][j] = 2
                        change = True
                    if grid[i][j] == 2 and j < len(grid)-1 and grid[i][j-1]==1:
                        grid[i][j-1] = 2
                        change = True
                    if grid[i][j] == 2 and j < len(grid)-1 and grid[i][j+1]==1:
                        grid[i][j+1] = 2
                        change = True

        if change:
            return count+1
        return count